File: React App/src/Task-5.js
// src/Task-5.js
import React, { useState, useEffect } from 'react';
import Plot from 'react-plotly.js';

const Task5 = () => {
    const [initialSpeed, setInitialSpeed] = useState(150);
    const [height, setHeight] = useState(300);
    const [gravity, setGravity] = useState(9.81);
    const [error, setError] = useState(null); // State to manage error messages
    const [data, setData] = useState([]); // State to store plot data

    const X = 1000;
    const Y = height; // Dynamically using height
    const u_given = initialSpeed; // Dynamically using initialSpeed

    const calculateMinimumLaunchSpeed = (X, Y) => {
        return Math.sqrt(gravity * (Y + Math.sqrt(X ** 2 + Y ** 2)));
    };

    const calculateLaunchAngles = (u, X, Y) => {
        const discriminant = u ** 4 - gravity * (gravity * X ** 2 + 2 * u ** 2 * Y);
        if (discriminant < 0) {
            throw new Error("No real solutions for the given parameters.");
        }
        const thetaHigh = Math.atan((u ** 2 + Math.sqrt(discriminant)) / (gravity * X));
        const thetaLow = Math.atan((u ** 2 - Math.sqrt(discriminant)) / (gravity * X));
        return [thetaHigh, thetaLow];
    };

    const calculateMinSpeedAngle = (X, Y) => Math.atan((Y + Math.sqrt(X ** 2 + Y ** 2)) / X);

    const generateTrajectory = (theta, v0, numPoints = 500) => {
        const tFlight = 2 * v0 * Math.sin(theta) / gravity;
        const t = Array.from({ length: numPoints }, (_, i) => (i / (numPoints - 1)) * tFlight);
        const x = t.map(ti => v0 * Math.cos(theta) * ti);
        const y = t.map(ti => v0 * Math.sin(theta) * ti - 0.5 * gravity * ti ** 2);
        return { x, y };
    };

    const generateBoundingParabola = (u, XMax, numPoints = 500) => {
        const x = Array.from({ length: numPoints }, (_, i) => (i / (numPoints - 1)) * XMax);
        const y = x.map(xi => (u ** 2 / (2 * gravity)) - (gravity / (2 * u ** 2)) * xi ** 2);
        return { x, y };
    };

    const calculateMaxRangeAngle = () => Math.PI / 4;

    useEffect(() => {
        try {
            const uMin = calculateMinimumLaunchSpeed(X, Y);
            const [thetaHigh, thetaLow] = calculateLaunchAngles(u_given, X, Y);
            const thetaMin = calculateMinSpeedAngle(X, Y);
            const thetaMaxRange = calculateMaxRangeAngle();

            const { x: xLow, y: yLow } = generateTrajectory(thetaLow, u_given);
            const { x: xHigh, y: yHigh } = generateTrajectory(thetaHigh, u_given);
            const { x: xMin, y: yMin } = generateTrajectory(thetaMin, uMin);
            const { x: xMaxRange, y: yMaxRange } = generateTrajectory(thetaMaxRange, u_given);
            const { x: xBound, y: yBound } = generateBoundingParabola(u_given, Math.max(...xMaxRange));

            setError(null); // Clear any previous errors
            setData([
                { x: xLow, y: yLow, type: 'scatter', mode: 'lines', line: { color: 'orange' }, name: 'Low ball' },
                { x: xHigh, y: yHigh, type: 'scatter', mode: 'lines', line: { color: 'blue' }, name: 'High ball' },
                { x: xMin, y: yMin, type: 'scatter', mode: 'lines', line: { color: 'gray' }, name: 'Min u' },
                { x: xMaxRange, y: yMaxRange, type: 'scatter', mode: 'lines', line: { color: 'red' }, name: 'Max range' },
                { x: xBound, y: yBound, type: 'scatter', mode: 'lines', line: { color: 'purple', dash: 'dash' }, name: 'Bounding parabola' },
                { x: [X], y: [Y], type: 'scatter', mode: 'markers', marker: { color: 'yellow', size: 10 }, name: 'Target (X, Y)' }
            ]);
        } catch (err) {
            setError(err.message); // Set error message if any issue occurs
            setData([]); // Clear data on error
        }
    }, [initialSpeed, height, gravity]);

    return (
        <div>
            <h1>Projectile to Hit (X, Y)</h1>
            {error ? (
                <div style={{ color: 'red' }}>
                    <h2>Error: {error}</h2>
                </div>
            ) : (
                <Plot
                    data={data}
                    layout={{
                        title: 'Projectile to Hit (X, Y)',
                        xaxis: { title: 'x / m' },
                        yaxis: { title: 'y above launch height / m' },
                    }}
                    style={{ width: '100%', height: '500px' }}
                />
            )}
            <div>
                <label>
                    Initial Speed (m/s):
                    <input
                        type="range"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={initialSpeed}
                        onChange={(e) => setInitialSpeed(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={initialSpeed.toFixed(2)}
                        onChange={(e) => setInitialSpeed(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
            <div>
                <label>
                    Initial Height (m):
                    <input
                        type="range"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={height}
                        onChange={(e) => setHeight(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={height.toFixed(2)}
                        onChange={(e) => setHeight(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
            <div>
                <label>
                    Gravity (m/s²):
                    <input
                        type="range"
                        min="0"
                        max="30"
                        step="0.01"
                        value={gravity}
                        onChange={(e) => setGravity(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="0"
                        max="30"
                        step="0.01"
                        value={gravity.toFixed(2)}
                        onChange={(e) => setGravity(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
        </div>
    );
};

export default Task5;

File: Tasks in Python/Task_5.py
import numpy as np
import matplotlib.pyplot as plt

g = 9.81
X = 1000
Y = 300
u_given = 150

def calculate_minimum_launch_speed(X, Y):
    u_min = np.sqrt(g * (Y + np.sqrt(X ** 2 + Y ** 2)))
    return u_min

def calculate_launch_angles(u, X, Y):
    discriminant = u ** 4 - g * (g * X ** 2 + 2 * u ** 2 * Y)
    if discriminant < 0:
        raise ValueError("No real solutions for the given parameters.")
    theta_high = np.arctan((u ** 2 + np.sqrt(discriminant)) / (g * X))
    theta_low = np.arctan((u ** 2 - np.sqrt(discriminant)) / (g * X))
    return theta_high, theta_low

def calculate_min_speed_angle(X, Y):
    theta_min = np.arctan((Y + np.sqrt(X ** 2 + Y ** 2)) / X)
    return theta_min

def generate_trajectory(theta, v0, num_points=500):
    t_flight = 2 * v0 * np.sin(theta) / g
    t = np.linspace(0, t_flight, num_points)
    x = v0 * np.cos(theta) * t
    y = v0 * np.sin(theta) * t - 0.5 * g * t ** 2
    return x, y

def generate_bounding_parabola(u, X_max, num_points=500):
    x = np.linspace(0, X_max, num_points)
    y = (u**2 / (2 * g)) - (g / (2 * u**2)) * x**2
    return x, y

def calculate_max_range_angle():
    return np.pi / 4

u_min = calculate_minimum_launch_speed(X, Y)
theta_high, theta_low = calculate_launch_angles(u_given, X, Y)
theta_min = calculate_min_speed_angle(X, Y)
theta_max_range = calculate_max_range_angle()

x_low, y_low = generate_trajectory(theta_low, u_given)
x_high, y_high = generate_trajectory(theta_high, u_given)
x_min, y_min = generate_trajectory(theta_min, u_min)
x_max_range, y_max_range = generate_trajectory(theta_max_range, u_given)

x_bound, y_bound = generate_bounding_parabola(u_given, max(x_max_range))

plt.figure(figsize=(10, 6))
plt.plot(x_low, y_low, label='Low ball', color='orange')
plt.plot(x_high, y_high, label='High ball', color='blue')
plt.plot(x_min, y_min, label='Min u', color='gray')
plt.plot(x_max_range, y_max_range, label='Max range', color='red')
plt.plot(x_bound, y_bound, label='Bounding parabola', color='purple', linestyle='dashed')
plt.scatter([X], [Y], color='yellow', label='Target (X,Y)', zorder=5)
plt.xlabel('x / m')
plt.ylabel('y above launch height / m')
plt.title('Projectile to hit (X,Y)')
plt.legend()
plt.grid()
plt.show()
